File: 2kyu/evaluate_mathematical_expression.hs
-- https://www.codewars.com/kata/52a78825cdfc2cfc87000005

import Prelude hiding (div)

import Data.Char (isDigit)
import Text.ParserCombinators.ReadP

nat :: ReadP String
nat = munch1 isDigit

number :: ReadP Double
number = do int  <- nat
            frac <- char '.' *> nat
            return $ read (int ++ '.' : frac)
     +++ do int <- nat
            return $ read int

ws :: ReadP String
ws = munch (`elem` " \t\n")

add = op '+' (+)
sub = op '-' (-)
mul = op '*' (*)
div = op '/' (/)

op c f = ws *> char c *> ws *> return f
     
brackets = char '(' *> ws *> expression <* ws <* char ')'

negated = (char '-' *> return negate) <*> factor

factor = number +++ brackets +++ negated
term   = mul +++ div
expr   = add +++ sub

expression :: ReadP Double
expression = factor `chainl1` term `chainl1` expr

calc :: String -> Double
calc = fst . head . readP_to_S (ws *> expression <* ws <* eof)


-- tests
test :: (String, Double) -> Either Double Bool
test (expr, answ)
  | actual == answ = Right True
  | otherwise      = Left  actual
  where actual = calc expr

main :: IO ()
main = do
  print $ test ("2 /2+3 * 4.75- -6", 21.25)
  print $ test ("2/(2+3)*4.33--6", 7.732)
  print $ test ("2 / (2 + 3) * 4.33 - -6", 7.732)
  print $ test (" \n(\t(\n86.63\t) *  \n\t-(\n64.2 )  +11.57\n\t*60.98\t /( 23.03\n) +\t(\n19.94\n*51.41 ) \n)+\t79.68 *-(\n\t20.74\t-\t \t59.4 * 47.64 \n+\n\t98.52 - -35.11 /( 96.87\t-\t5.72\t)\n\n/(\t19.48\n) - 30.65 + 57.27\n   ) \t\n/(\t(\n58.83\n/( (\t1.08\n))\n) )\t/( (\t79.22\t- \t58.61\t- \t50.65+\n61.78 \t *\n\n13.03* 36.55\n \n-\t 19.94\n+ \n 45.5+\t0.6 -  49.1 +\n\t71.05\n/(\n--82.75\t\t-76.04\n+ 23.1 * (\n68.57\t)\n- \t -(\t  (\n66.99 )\n/( \t59.32\t*--13.05\t*\t92.86 \n\n)\n-\n--43.8 \n * 13.93\t\n)\t\t-\t57.03+ \n(\t43.81 )\n+ 64.68\n\n/(\n85.94 )\t\t-\t\n4.36*\t 24.53/( 80.2\t) \t  )\t\n\n/(\n24.11\n) \t\n)\t)\n", -4505.761294455091)

File: 2kyu/evaluate_mathematical_expression.py
# https://www.codewars.com/kata/52a78825cdfc2cfc87000005

import re
from operator import add, sub, mul, truediv as div

OPS = {
    '+': add, 
    '-': sub,
    '*': mul,
    '/': div
      } 

tokenizer = re.compile(r'[.\d]+|[()\+\-\*\/]')

def calc(expr):
    tokens = tokenizer.findall(expr)
    res, _ = expression(tokens, 0)
    return res
    
def expression(tokens, id):
    value, id = term(tokens, id)

    while id < len(tokens) and tokens[id] != ')':
        tok = tokens[id]
        if tok in '+-':
            next, id = term(tokens, id+1)
            value = OPS[tok](value, next)

    return value, id

def term(tokens, id):
    value, id = factor(tokens, id)

    while id < len(tokens) and tokens[id] in '*/':
        op = tokens[id]
        next, id = factor(tokens, id+1)
        value = OPS[op](value, next)

    return value, id

def factor(tokens, id):
    tok = tokens[id]

    if tok == '(':
        value, id = expression(tokens, id+1)
    elif tok == '-':
        value, id = factor(tokens, id+1)
        value, id = -value, id-1
    else:
        value = float(tok)

    return value, id+1
    
### TESTS
if __name__ == '__main__':
    TESTS = [("1 + 1", 2),
             ("8/16", 0.5),
             ("2 + 2 + 2 - 2", 4),
             ("2 * 2 + 2", 6),
             ("2 + 2 - 2", 2),
             ("(2 + 3) * 2 + 7 * 3", 31),
             ("(2 + 2) * 2", 8),
             ("-(2 + 2) * 2", 8),
             ("1.5 * -1.5", -2.25),
             ("3 -(-1)", 4),
             ("2 + -2", 0),
             ("10- 2- -5", 13),
             ("(((10)))", 10),
             ("-7 * -(6 / 3)", 14)
             ]
    for exp, ans in TESTS:
        actual = calc(exp)
        if actual == ans:
            print(f"{exp} ok")
        else:
            print(f"{exp} expected: {ans}, but was: {actual}")
